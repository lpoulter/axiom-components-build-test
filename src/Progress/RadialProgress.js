import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Base from '../Base/Base';
import './RadialProgress.css';

export default class RadialProgress extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string.isRequired,
    diameter: PropTypes.number.isRequired,
    size: PropTypes.string,
    sizeRem: PropTypes.string,
  };

  render() {
    const { children, className, diameter, size, sizeRem, ...rest } = this.props;
    const style = {};
    const classes = classnames(className, 'ax-radial-progress', {
      [`ax-radial-progress--${size}`]: size,
    });

    if (sizeRem) {
      style.width = sizeRem;
      style.height = sizeRem;
    }

    return (
      <Base { ...rest }
          Component="svg"
          className={ classes }
          style={ style }
          viewBox={ `0 0 ${ diameter } ${ diameter }` }>
        <g transform={ `translate(${ diameter / 2 }, ${ diameter / 2 }) rotate(-90)` }>
          { children }
        </g>
      </Base>
    );
  }
}
