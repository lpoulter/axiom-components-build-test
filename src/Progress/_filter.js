import { svgDefineOnce } from '@brandwatch/axiom-utils';

export default () => {
  ['subtle', 'white'].forEach(color => {
    svgDefineOnce(`
      <linearGradient id="ax-progress-infinite__filter--${ color }">
        <stop class="ax-progress-infinite__arc-gradient-stop--${ color }" offset="0%"></stop>
        <stop class="ax-progress-infinite__arc-gradient-stop--${ color }" offset="50%"></stop>
        <stop class="ax-progress-infinite__arc-gradient-stop--${ color }" offset="90%"></stop>
      </linearGradient>
    `);
  });
};
