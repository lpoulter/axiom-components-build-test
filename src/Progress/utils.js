const DIAMETER = 24;
const STROKE_WIDTH = 4;
const INNER_RADIUS = DIAMETER / 2 - STROKE_WIDTH / 2;
const STROKE_DASH_ARRAY = 2 * Math.PI * INNER_RADIUS;

export const getArcProperties = ({ percent }) => ({
  diameter: DIAMETER,
  innerRadius: INNER_RADIUS,
  strokeWidth: STROKE_WIDTH,
  strokeDasharray: STROKE_DASH_ARRAY,
  strokeDashoffset: ((100 - percent) / 100) * STROKE_DASH_ARRAY,
});
