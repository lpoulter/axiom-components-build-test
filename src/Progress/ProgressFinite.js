import PropTypes from 'prop-types';
import React, { Component } from 'react';
import renderFilter from './_filter';
import { getArcProperties } from './utils';
import RadialProgress from './RadialProgress';
import './ProgressFinite.css';

export default class ProgressFinite extends Component {
  static propTypes = {
    /** Percentage of progress complete */
    percent: PropTypes.number.isRequired,
    /** Size of the indicator */
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    /** Size in REM units */
    sizeRem: PropTypes.string,
  };

  static defaultProps = {
    percent: 0,
    size: 'small',
  }

  UNSAFE_componentWillMount = renderFilter

  render() {
    const {
      percent,
      ...rest
    } = this.props;

    const {
      diameter,
      innerRadius,
      strokeWidth,
      strokeDasharray,
      strokeDashoffset,
    } = getArcProperties({ percent });

    return (
      <RadialProgress { ...rest } className="ax-progress-finite" diameter={ diameter }>
        <circle
            className="ax-progress-finite__background"
            r={ innerRadius }
            style={ { strokeWidth } } />
        <circle
            className="ax-progress-finite__arc"
            r={ innerRadius }
            style={ { strokeDasharray, strokeDashoffset, strokeWidth } } />
      </RadialProgress>
    );
  }
}
