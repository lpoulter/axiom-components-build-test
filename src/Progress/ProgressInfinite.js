import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import { getArcProperties } from './utils';
import renderFilter from './_filter';
import RadialProgress from './RadialProgress';
import './ProgressInfinite.css';

export default class ProgressInfinite extends Component {
  static propTypes = {
    /** Color of the indicator */
    color: PropTypes.oneOf([
      'subtle',
      'white',
    ]),
    /** Size of the indicator */
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    /** Size in REM units */
    sizeRem: PropTypes.string,
  };

  static defaultProps = {
    color: 'subtle',
    size: 'small',
  }

  UNSAFE_componentWillMount = renderFilter

  render() {
    const {
      color,
      ...rest
    } = this.props;
    const classes = classnames('ax-progress-infinite', `ax-progress-infinite--${color}`);
    const {
      diameter,
      innerRadius,
      strokeWidth,
      strokeDasharray,
      strokeDashoffset,
    } = getArcProperties({ percent: 75 });

    return (
      <RadialProgress { ...rest } className={ classes } diameter={ diameter }>
        <circle
            className="ax-progress-infinite__arc"
            r={ innerRadius }
            style={ {
              stroke: `url(#ax-progress-infinite__filter--${ color })`,
              strokeDasharray,
              strokeDashoffset,
              strokeWidth,
            } } />
      </RadialProgress>
    );
  }
}
