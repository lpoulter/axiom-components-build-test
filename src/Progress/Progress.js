import PropTypes from 'prop-types';
import React, { Component } from 'react';
import AlertIcon from '../AlertIcon/AlertIcon';
import ProgressFinite from './ProgressFinite';
import ProgressInfinite from './ProgressInfinite';

export default class Progress extends Component {
  static propTypes = {
    /** Shows a successful completion state */
    complete: PropTypes.bool,
    /** Shows an error state */
    error: PropTypes.bool,
    /** Percentage of progress */
    percent: PropTypes.number,
    /** Size of the indicator */
    size: PropTypes.oneOf(['small', 'medium', 'large']),
    /** Size in REM units */
    sizeRem: PropTypes.string,
  };

  static defaultProps = {
    size: 'small',
  };

  render() {
    const {
      complete,
      error,
      percent,
      ...rest
    } = this.props;

    if (error) {
      return <AlertIcon { ...rest } type="error" />;
    }

    if (complete) {
      return <AlertIcon { ...rest } type="success" />;
    }

    if (!isNaN(parseFloat(percent))) {
      return <ProgressFinite { ...rest } percent={ percent } />;
    }

    return <ProgressInfinite { ...rest } />;
  }
}
