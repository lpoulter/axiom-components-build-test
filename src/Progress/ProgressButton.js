import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Button from '../Button/Button';
import Cloak from '../Cloak/Cloak';
import ProgressInfinite from './ProgressInfinite';
import './ProgressButton.css';

const progressSizeMap = {
  small: '1rem',
  medium: '1.25rem',
  large: '1.5rem',
};

export default class ProgressButton extends Component {
  static propTypes = {
    /** Content to be inserted into the Button */
    children: PropTypes.node.isRequired,
    /** Flag if an action is in progress, shows the progress indicator */
    isInProgress: PropTypes.bool,
    /** Size of the Button. See Button[size]. */
    size: PropTypes.oneOf(['small', 'medium', 'large']),
  };

  render() {
    const {
      children,
      isInProgress,
      size = Button.defaultProps.size,
      ...rest
    } = this.props;

    const classes = classnames('ax-progress-button', {
      'ax-progress-button--in-progress': isInProgress,
    });

    return (
      <Button { ...rest }
          active={ isInProgress }
          size={ size }
          style="primary">
        <div className={ classes }>
          <Cloak
              className="ax-progress-button__content"
              invisible={ isInProgress }>
            { children }
          </Cloak>

          <Cloak
              className="ax-progress-button__indicator"
              invisible={ !isInProgress }>
            <ProgressInfinite
                color="white"
                sizeRem={ progressSizeMap[size] } />
          </Cloak>
        </div>
      </Button>
    );
  }
}
