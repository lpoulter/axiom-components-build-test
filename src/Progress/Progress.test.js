import React from 'react';
import renderer from 'react-test-renderer';
import Progress from './Progress';

const getComponent = (props = {}) =>
  renderer.create(
    <Progress { ...props } />
  );

describe('Progress', () => {
  it('renders with defaultProps', () => {
    const component = getComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an error indicator', () => {
    const component = getComponent({ error: true });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a success indicator', () => {
    const component = getComponent({ complete: true });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a finite progress indicator', () => {
    const component = getComponent({ percent: 50 });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
