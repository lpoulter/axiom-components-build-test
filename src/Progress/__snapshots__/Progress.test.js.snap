// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress renders a finite progress indicator 1`] = `
<svg
  className="ax-progress-finite ax-radial-progress ax-radial-progress--small"
  style={Object {}}
  viewBox="0 0 24 24"
>
  <g
    transform="translate(12, 12) rotate(-90)"
  >
    <circle
      className="ax-progress-finite__background"
      r={10}
      style={
        Object {
          "strokeWidth": 4,
        }
      }
    />
    <circle
      className="ax-progress-finite__arc"
      r={10}
      style={
        Object {
          "strokeDasharray": 62.83185307179586,
          "strokeDashoffset": 31.41592653589793,
          "strokeWidth": 4,
        }
      }
    />
  </g>
</svg>
`;

exports[`Progress renders a success indicator 1`] = `
<div
  className="ax-alert-icon ax-alert-icon--primary ax-alert-icon--small ax-alert-icon--success"
>
  <svg
    className="ax-icon ax-icon--tick"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    style={
      Object {
        "height": "1rem",
        "width": "1rem",
      }
    }
    viewBox="0 0 16 16"
  />
</div>
`;

exports[`Progress renders an error indicator 1`] = `
<div
  className="ax-alert-icon ax-alert-icon--primary ax-alert-icon--small ax-alert-icon--error"
>
  <svg
    className="ax-icon ax-icon--warning"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    style={
      Object {
        "height": "1rem",
        "width": "1rem",
      }
    }
    viewBox="0 0 16 16"
  />
</div>
`;

exports[`Progress renders with defaultProps 1`] = `
<svg
  className="ax-progress-infinite ax-progress-infinite--subtle ax-radial-progress ax-radial-progress--small"
  style={Object {}}
  viewBox="0 0 24 24"
>
  <g
    transform="translate(12, 12) rotate(-90)"
  >
    <circle
      className="ax-progress-infinite__arc"
      r={10}
      style={
        Object {
          "stroke": "url(#ax-progress-infinite__filter--subtle)",
          "strokeDasharray": 62.83185307179586,
          "strokeDashoffset": 15.707963267948966,
          "strokeWidth": 4,
        }
      }
    />
  </g>
</svg>
`;
